### PR TITLE
make timeout for brute force operation configurable

### LINKED ISSUE #3276: Add additional annotation that allows specifying the bruteforce protection timeout
As stated at https://github.com/nextcloud/server/pull/3265#issuecomment-275362713

### PR DESCRIPTION
**Pull request by schiessle**
General default is 12 hours, default for API calls is 10 minutes. If you want to adjust it for your API call use:

````
/**
 * @BruteForceProtection <myapi>
 * @BruteForceProtectionTimeout <seconds>
 */
````

fix https://github.com/nextcloud/server/issues/3276

### CODE REVIEW

File: tests/lib/AppFramework/Middleware/Security/SecurityMiddlewareTest.php [hunk 1]
```diff
@@ -723,4 +729,48 @@ public function dataTestBeforeControllerBruteForce() {
...
+
+		$this->assertSame($expected, $this->invokePrivate($middleware, 'getTimeout'));
+	}
+
+	public function dataTestGetTimeout() {
+		return [
+			[true, 60, 60],
+			[true, '60', 60],
+			[true, 'foo', 600],
+			[false, 60, 600],
```
Comment (nickvergessen) [inline tests/lib/AppFramework/Middleware/Security/SecurityMiddlewareTest.php]: test float, double and `int.string`

**Review by nickvergessen — CHANGES_REQUESTED** [review]

File: tests/lib/AppFramework/Middleware/Security/SecurityMiddlewareTest.php [hunk 2]
```diff
@@ -723,4 +729,52 @@ public function dataTestBeforeControllerBruteForce() {
...
+
+		$this->assertSame($expected, $this->invokePrivate($middleware, 'getTimeout'));
+	}
+
+	public function dataTestGetTimeout() {
+		return [
+			[true, 60, 60],
+			[true, '60', 60],
+			[true, '60.3', 600],
+			[true, 60.3, 600],
```
Comment (MorrisJobke) [inline tests/lib/AppFramework/Middleware/Security/SecurityMiddlewareTest.php]: I don't get this. Why should 60.3 create a timeout of 600?

Comment (MorrisJobke) [inline tests/lib/AppFramework/Middleware/Security/SecurityMiddlewareTest.php]: Ah - it's the default

**Comment by blizzz** [comment]
Read through the connected issues, but it is not clear to me where this benefits us? Why would someone set different values?

**Comment by blizzz** [comment]
from IRC:

> <BernhardPosselt> i think thats the wrong way to do it
> <BernhardPosselt> apart from that you will break compat with doctrine annotations (if you ever want to go that way)
> <BernhardPosselt> if it should be configurable, then put it into a section inside the config.php
> <BernhardPosselt> one that applies to every request because brute force protection protects against password guessing, so if you add a smaller period you will weaken everything

**Comment by schiessle** [comment]
> Read through the connected issues, but it is not clear to me where this benefits us? Why would someone set different values?

In the annotations we usually use it to protect against public API calls, e.g. if people send you a federated share. There we want to avoid that people can spam you with hundreds of federated shares, etc. in a few minutes. Therefore it makes sense to have a shorter timespan here.

**Comment by BernhardPosselt** [comment]
Sounds to me like API rate limiting which is AFAIK not yet built in and serves a different usecase than brute force protection

**Comment by BernhardPosselt** [comment]
Because brute force protection only works on failed login attempts, correct @LukasReschke ?

**Comment by nickvergessen** [comment]
That is why we build the API, so it can be on other things too, not only on failed login attempts...

**Comment by BernhardPosselt** [comment]
I see, then you probably want to rename the annotations to something like 

    @RateLimit(1000)

which would limit the number of requests per caller to maybe 1000 per hour.

**Comment by BernhardPosselt** [comment]
Check out http://www.django-rest-framework.org/api-guide/throttling/ for more inspiration ;D (that's what we use for the app store)

**Comment by LukasReschke** [comment]
I have been thinking whether it would make sense to add a function `increaseThrottler` to `\OCP\AppFramework\Controller`. That way in the controller you'd have to call:

`$this->increaseThrottler()` to increase the counter, this is especially useful for the case where we want to use this annotation for a bruteforce protection. The function would basically just increase the amount, the sleep would still be done by the annotation.

For the use-case of API rate limiting one would just have to add that at the start of the function. That would give us way more flexibility here. Also that will solve the problem with https://github.com/nextcloud/server/pull/3265 as we could simply only increase the counter when the password is wrong.

Thoughts?

**Comment by LukasReschke** [comment]
Or we could go with a `@RateLimit(1000)` one which automatically counts. And a separate one `@BruteforceProtection` one which needs to call `$this->increaseThrottler()`.

**Comment by nickvergessen** [comment]
Sounds good to me

**Comment by LukasReschke** [comment]
Replaced by https://github.com/nextcloud/server/pull/4336

**Pull request closed without merging**